src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -49,38 +73,91 @@ public abstract class AssemblyLoadContext
         internal static extern void InternalStartProfile(string profile, IntPtr ptrNativeAssemblyLoadContext);
 #endif // FEATURE_MULTICOREJIT
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppContext.Unloading
+            AppContext.Unloading += OnAppContextUnloading;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+#if FEATURE_COLLECTIBLE_ALC
+        protected AssemblyLoadContext() : this(false, false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
-        
-        void InitializeLoadContext(bool fRepresentsTPALoadContext)
+
+        protected AssemblyLoadContext(bool isCollectible) : this(false, isCollectible)
```

Comment:
Yes looks like these new public/protected APIs would need to be there regardless. Would be good to have one definition  for the method/property and have a different implementation in the body for when the feature is defined or not defined. `InvalidOperationException` is already what you're using in Unload when the instance is not collectible, I think that would work for the case where the feature is disabled as well.